Mathlib/Order/CompletePartialOrder.lean

```diff
@@ -0,0 +1,243 @@
+/-
+Copyright (c) 2023 Christopher Hoskin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Christopher Hoskin
+-/
+import Mathlib.Order.Directed
+import Mathlib.Order.OmegaCompletePartialOrder
+import Mathlib.Data.Set.Finite
+import Mathlib.Data.Finset.Basic
+
+/-!
+# Complete Partial Order
+
+This file begins by showing that each set `s` in a join-semilattice generates a directed set `d`
+such that:
+
+- `s` is a subset of `d`
+- `s` and `d` share the same upper bounds
+- `u` is the least upper bound of `s` if and only if it is the least upper bound of `d`
+
+It follows that if every directed set in a join-semilattice has a least upper bound then the join is
+complete.
+
+The second part of this file considers complete partial orders (sometimes called directedly complete
+partial orders). These are partial orders for which every directed set has a least upper bound.
+
+## References
+
+- [B. A. Davey and H. A. Priestley, Introduction to lattices and order][davey_priestley]
+
+## Tags
+
+complete partial order, directedly complete partial order
+
+-/
+
+section SemilatticeSup
+
+structure DirectedSet (α : Type _) [PartialOrder α] where
+  set : Set α
+  directed : DirectedOn (. ≤ .) set
```

Comment:
Will you ever consider operations on the type of directed sets? So far you don't, and if you are not going to later, you should not define `DirectedSet` and instead use `Set` directly.

Reply:
@YaelDillies I was basing this on `Order.OmegaCompletePartialOrder` since this seems to be the non-countable analogue.

Reply:
Okay, I've removed `DirectedSet` and used `Set` instead.